compiler/rustc_typeck/src/check/demand.rs

```diff
@@ -356,9 +356,36 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 })
                 .collect();
 
-            if let [variant] = &compatible_variants[..] {
+            if self.is_hir_id_from_struct_pattern_shorthand_field(expr.hir_id, expr.span) {
+                if let Ok(code) = self.tcx.sess.source_map().span_to_snippet(expr.span) {
+                    if let [variant] = &compatible_variants[..] {
```

Comment:
Maybe make this a match instead, like this?

```rust
match &compatible_variants[..] {
  [] => /* can't format this */
  [variant] => ...,
  _ => ...,
}
```

Reply:
Sure.  I was matching the style of the code I split out into this case, but that's not hard to fix as a drive-by.